### ISSUE TITLE
User applications don't launch / scale correctly after restore

### ISSUE THREAD

**Issue by temujin9** [flynn/flynn#3382]
Following a replacement via restore from backup, two different clusters failed to launch containers for the same application code ("firebird-firebuild"). In both, a push to the application in question restored functionality. 

In https://bazaarvoice.box.com/s/7sskqao1v0r8xmnm7oo28ks51cluyzbd, the push left the old release of qa-firebird-firebuild and its containers scaled up, although those containers are now logging nothing new (`flynn log -j XXX -f` sits idle). That may be symptomatic of a separate problem.

In https://bazaarvoice.box.com/s/85nvgzy8vu90oios6luqbr8y7rc51ob7, it scaled down the old containers for prod-firebird-firebuild as expected. However, staging-firebird-firebuild had the same old release issue that qa-firebird-firebuild did, save that both old and new processes are processing correctly (i.e. logging output).

(Please ignore the "scaler" activity. It is a red herring: a noop process intended as a stub for later development, which accidentally got left on (and thus, flapping), but which should have no effect otherwise. Also ignore the bare "firebird-firebuild" in the first cluster: it was a vestigial app from early in their dev process, and never got a release.)

**Comment by temujin9** [comment]
Correction: The containers for old releases launched, and ran for a little while, then stopped logging (and presumably, doing anything) for roughly 5-8 hrs. After a push, both old (if retained) and new containers run correctly.

**Comment by temujin9** [comment]
https://botbot.me/freenode/flynn/2016-09-13/?msg=72985067&page=2 has the initial debugging conversation with @lmars.

**Comment by lmars** [comment]
After the push, the old jobs remained running because their state was not persisted to the controller due to a database timeout error, for example:

```
t=2016-09-12T17:35:01+0000 lvl=info msg="request started" component=controller req_id=09d69f04-48a2-48ef-94b0-846e5257e573 method=PUT path=/apps/c80ddfb6-6044-4276-b08f-ed01355d4d5f/jobs/7f2044b8-f384-4acb-8f9f-bf7a3112bd07 client_ip=100.100.75.6
t=2016-09-12T17:35:32+0000 lvl=eror msg="Timeout: All connections in pool are busy" component=controller req_id=09d69f04-48a2-48ef-94b0-846e5257e573
t=2016-09-12T17:35:32+0000 lvl=info msg="request completed" component=controller req_id=09d69f04-48a2-48ef-94b0-846e5257e573 status=500 duration=30.297764984s
```

This timeout came just after the cluster had booted and there was a lot of activity, the timeouts stopped after another couple of minutes.

Since the returned error was not a retryable one, the scheduler just accepted the failure and did not try to persist the job state again. This bug in the scheduler has now been fixed in #3384.

As for the behaviour before the push, I looked at the logs for the QA cluster, and was provided with the full log for job `ip1010036255-7f2044b8-f384-4acb-8f9f-bf7a3112bd07`. The job started at `2016-09-12T17:35:08Z` (which is when the cluster was restored) and contains log lines up until `2016-09-12T18:59:13Z`, then there is a gap until `2016-09-13T02:53:53Z`, which is also when a deployment occurred.

There doesn't seem to be much Flynn could have done in this situation (it is not aware of what jobs should be doing, just that they should be running). My suggestion is to add a HTTP health check to these worker processes which checks that they are functioning correctly and specify that in the release so that Flynn will kill the job if the health check fails. Let us know if you need more information on getting that working.
